src/core/lwt.ml

```diff
@@ -455,10 +455,11 @@ struct
   and 'a resolved_state = ('a, underlying, resolved) state
 
   and how_to_cancel =
-    | Not_cancelable              :                           how_to_cancel
-    | Cancel_this_promise         :                           how_to_cancel
-    | Propagate_cancel_to_one     : (_, _, _) promise      -> how_to_cancel
-    | Propagate_cancel_to_several : (_, _, _) promise list -> how_to_cancel
+    | Not_cancelable                           :                           how_to_cancel
+    | Cancel_this_promise                      :                           how_to_cancel
+    | Cancel_this_promise_and_propagate_to_one : (_, _, _) promise      -> how_to_cancel
```

Comment:
Why is it necessary to add a new constructor? Given you were able to implement this function outside Lwt, I imagine it should be possible to implement it inside without changing these constructors.

Note that the existing constructors have a certain "algebraic" nature to them, in that `Not_cancelable` and `Cancel_this_promise` correspond to zero cancelable "ancestors," and arise when entirely independent promises are introduced ex nihilo with `Lwt.task` and `Lwt.wait` (functions that take zero other promise arguments), `Propagate_cancel_to_one` corresponds to one cancelable "ancestor," and arises from sequential composition upon one existing promise, and `Propagate_cancel_to_several` corresponds to an arbitrary number of cancelable "ancestors," and arises from concurrent composition upon an arbitrary number of existing promises. So these existing constructors really are in some way basic, they are a kind of 0, 1, and *.

Note also that `protected` and `no_cancel` are implemented without introducing new constructors.

Reply:
Oh, I can do a `Propagate_cancel_to_several` and pass it “itself” I guess. I'll do that instead.

Reply:
I had to use `lazy` to fit in the recursion. But it seems to work. At least the tests written for it pass.